### ISSUE TITLE
Inspect timing-out aead tests

### ISSUE THREAD

**Issue by th4s** [tlsnotary/tlsn#548]
PR #547 comments the aead tests, which are currently timing out. This needs further investigation as it is not locally reproducible.

**Comment by th4s** [comment]
PR #558 is used for investigating.

**Comment by th4s** [comment]
This deadlock is hard to fix. Things I tried so far:

- I could not reproduce this bug on my local machine, even when using act and running it in docker.
- Deadlock seems to vanish when executing a single AEAD test. It is strange that either all 4 of the started aead tests hang or all pass. I never observed some test passing and another hanging.
- Looking for `static` variables, i.e. shared state between unit tests -> nothing found
- Manual code inspection, especially of `mpz-garble/deap`-> nothing found so far. 
- Replacing all `.lock().unwrap()` with `try_lock().unwrap()` -> deadlock vanishes.
- Adding tracing.
  - There seems to be a problem with [`set_default`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/util/trait.SubscriberInitExt.html#method.set_default) in that it omits some logging statements. Couldn't make any sense out of it, will probably create an issue for this. However sometimes I was able to capture a little bit of output. This can be found in the failed runs in [these PRs](https://github.com/th4s/tlsn/pulls)
  - When trying to add proper tracing via [`set_global_default`](https://docs.rs/tracing/latest/tracing/subscriber/fn.set_global_default.html) the deadlock vanishes.
- Removing tracing and using `println!` statements the deadlock vanishes.

Saying the deadlock vanishes means that I tried to run it several times (4-5) and didn't encounter it.

**Comment by th4s** [comment]
Making some progress by carefully adding `println!` statements. This run is interesting: https://github.com/th4s/tlsn/actions/runs/10506339441/job/29106915720
It hangs in `decrypt_private -> verify_tag -> compute_tag_share -> share_keystream_block -> compute_keystream -> compute -> execute`

**Comment by th4s** [comment]
New hanging run with more logging: https://github.com/th4s/tlsn/actions/runs/10507218244/job/29108495136
It hangs inside `generate` (also in `evaluate` but this is probably just a consequence of waiting for `generate`).

My current bet would be that it has to do with `feed, flush` and the limited backpressure of the `TestSTExecutor`.

**Comment by th4s** [comment]
New hanging run with more fine-grained logging: https://github.com/th4s/tlsn/actions/runs/10509969553/job/29118455831
 
Seems to be hanging here: https://github.com/privacy-scaling-explorations/mpz/blob/7783045420272defb0caf52c0b7ceeab03badb55/crates/mpz-garble/src/generator/mod.rs#L309-L318

**Comment by th4s** [comment]
This run https://github.com/th4s/tlsn/actions/runs/10523156760 provides evidence that it hangs on the evaluator side here:
https://github.com/privacy-scaling-explorations/mpz/blob/8a6ae620fc67176111935a8b1ac8b10c63d8e253/crates/mpz-garble/src/evaluator/mod.rs#L393-L470

**Comment by th4s** [comment]
When running tests from the workspace root, the feature `rayon` although not specified, is activated by `tlsn-verifier` and `tlsn-prover`.

**Comment by th4s** [comment]
What happens is the following:
1) When running our test-suite from the workspace root, `tlsn-verifier/tlsn-prover` activate `rayon` for `mpz-common`. This means that although not specified, `tlsn-aead` unit tests are run with this feature enabled.
2) Rayon uses `std::thread::available_parallelism` to estimate the amount of parallelism (when the rayon threadpool is left unconfigured as we do), i.e. how many tasks are run in parallel in the rayon threadpool. I determined this to be `4` for our github CI.
3) When running unit tests, Rust uses 4 threads to parallelize test execution. But in each aead test the leader **AND** and the follower make use of `rayon::spawn`.
5) All the 4 rayon-spawned follower tasks use `expect_next` in `evaluate` to get garbling batches from the generator. This means that when all 4 follower tasks are running in rayon, the rayon threadpool is saturated. The generator can only queue their rayon tasks to send garbling batches but they are never executed because the follower tasks will never finish (since they are waiting for garbling batches) -> deadlock

**Comment by th4s** [comment]
Yeah I think I am leaning towards increasing the thread count for the aead unit tests.

**Comment by heeckhau** [comment]
Fixed in #575
